pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -468,19 +488,179 @@ func (m panicNodeMutator) Generate(
 		addPanicStep := stepToPanic.
 			InsertBefore(panicNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode})
 		var addRestartStep []mutation
+		var restartStep stepSelector
 		// If validEndStep is nil, it means that there are no steps after the panic step that
 		// are compatible with a dead node, so we immediately restart the node after the panic.
 		if validEndStep == nil {
+			restartStep = cutStep
 			addRestartStep = cutStep.InsertBefore(restartNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt, restartDesc})
 		} else {
-			addRestartStep = validEndStep.
-				RandomStep(rng).
+			restartStep = validEndStep.RandomStep(rng)
+			addRestartStep = restartStep.
 				Insert(rng, restartNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt, restartDesc})
 		}
 
+		failureContextSteps, _ := validStartStep.CutBefore(func(s *singleStep) bool {
+			return s == restartStep[0]
+		})
+		failureContextSteps.MarkInFailureContext()
+
 		mutations = append(mutations, addPanicStep...)
 		mutations = append(mutations, addRestartStep...)
 	}
 
-	return mutations
+	return mutations, nil
+}
+
+func (m panicNodeMutator) SupportedDeployments() map[DeploymentMode]struct{} {
+	return map[DeploymentMode]struct{}{
+		SharedProcessDeployment:   {},
+		SystemOnlyDeployment:      {},
+		SeparateProcessDeployment: {},
+	}
+}
+
+type networkPartitionMutator struct{}
+
+func (m networkPartitionMutator) Name() string { return failures.IPTablesNetworkPartitionName }
+
+func (m networkPartitionMutator) Probability() float64 {
+	return 0.3
+}
+
+func (m networkPartitionMutator) Generate(
+	rng *rand.Rand, plan *TestPlan, planner *testPlanner,
+) ([]mutation, error) {
+	var mutations []mutation
+	upgrades := randomUpgrades(rng, plan)
+	idx := newStepIndex(plan)
+	nodeList := planner.currentContext.System.Descriptor.Nodes
+
+	for _, upgrade := range upgrades {
+		possiblePointsInTime := upgrade.
+			Filter(func(s *singleStep) bool {
+				// We don't want to set up a partition concurrently with other steps, and inserting
+				// before a concurrent step causes the step to run concurrently with that step, so
+				// we filter out any concurrent steps. We also filter out any steps that are
+				// already in the context of a failure to avoid overlapping failure injections.
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s) && !s.inFailureContext
+			})
+
+		stepToPartition := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		var firstStepInConcurrentBlock *singleStep
+
+		isInvalidRecoverStep := func(s *singleStep) bool {
+			// Restarting a node in the middle of a network partition has a chance of
```

Comment:
Only if the partition with quorum were to lose this node. Are there other cases?